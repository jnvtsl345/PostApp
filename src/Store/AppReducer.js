function userReducer(state,action){
    switch(action.type) {
        case 'LOGIN':
            state.username = action.userDetails.username;
            state.password = action.userDetails.password;
            return state;
        case 'LOGOUT':
            return {
                username : '',
               password : ''
            }
        default:
            return state;       
    }
}

function postsReducer(state,action) {
    switch (action.type){
        case 'CREATE_POST':
             return [action.post,...state];   
        default:
            return state;
    }
}

export default function appReducer(state,action){
    return{
        user: userReducer(state.user, action),
        posts: postsReducer(state.posts, action)
    }
}