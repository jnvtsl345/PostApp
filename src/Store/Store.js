import React,{ useReducer } from 'react';
import { createContext } from 'react';
import appReducer from './AppReducer';
export const GlobalContext = createContext({
    state: {},
    dispatch: {}
})

export const Store= (props) =>{
    const initialState ={
        user:{username:'',password:''},
        posts:[{title: 'React', content: 'A JavaScript library for building user interfaces', author: 'someuser'},
               {title: 'React Hooks', content: 'JavaScript functions to hook into React state', author: 'someuser'},
               {title: 'Material UI', content: 'A popular React UI Framework', author: 'someuser'}
            ]
    } 
    
    const [state,dispatch] = useReducer(appReducer,initialState);
      return(
          <GlobalContext.Provider value={{state,dispatch}} >
              {props.children}
          </GlobalContext.Provider>
    )
}