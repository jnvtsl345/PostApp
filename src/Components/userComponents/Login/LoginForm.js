import React, {useState, useContext} from 'react' ;
import loginStyle from'./LoginForm.module.css';
import { GlobalContext } from '../../../Store/Store';

function LoginForm(){
  //Using Context to get Dispatcher and Set Local State
  const {dispatch} = useContext(GlobalContext);
  const [userDetails,setDetails] = useState({username:"",password:""});
  
  //Submit Handler for Login Form
  const submitHandler = e =>{
    e.preventDefault();
    if(userDetails.username.length ===0 || userDetails.password.length ===0 ){
      alert("Username and Password cannot be empty!!");
    }
    else{
      dispatch({type:"LOGIN",userDetails});
      setDetails({username:"",password:""});
    }
  }
  
  //React Controlled Login Form
  return(
    <form data-testid="loginForm" onSubmit={submitHandler} className={loginStyle.loginForm}>
      <label htmlFor="username">Username:</label>
      <input type="text" name="username" id="username" value={userDetails.username} data-testid="username" onChange={e => setDetails({...userDetails, username: e.target.value})} required />
      <label htmlFor="username">Password:</label>
      <input type="password" name="password" id="password" value={userDetails.password} data-testid="password" onChange={e => setDetails({...userDetails, password: e.target.value})} required />
      <input type="submit" data-testid="btnLogin" value="Login" />
    </form>
  ) 
}

export default LoginForm;