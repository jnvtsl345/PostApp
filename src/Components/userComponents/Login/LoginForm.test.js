import React from 'react';
import ReactDom from 'react-dom';
import LoginForm from './LoginForm.js';
import { render, cleanup, fireEvent, getByTestId} from '@testing-library/react';
import { Store } from '../../../Store/Store';
import "@testing-library/jest-dom/extend-expect";
import renderer from 'react-test-renderer';

afterEach(cleanup);
//Checking if it renders correctly
it("renders component without crashing", () =>{
    const div =document.createElement("div");
    ReactDom.render(<LoginForm ></LoginForm>,div);
});

//Snapshot Testing
it("matches snapshot",()=>{
    const tree = renderer.create(<LoginForm></LoginForm>).toJSON();
    expect(tree).toMatchSnapshot();
})

//Checking if our inputs are storing values or not
it("Input's Work",()=>{
    const {getByTestId} = render(<Store><LoginForm></LoginForm></Store>)
    fireEvent.change(getByTestId("username"),{target:{value:'vatsal'}})
    fireEvent.change(getByTestId("password"),{target:{value:'vatsal345'}})
    expect(getByTestId("username").value).toBe('vatsal')
    expect(getByTestId("password").value).toBe('vatsal345')
})

//Checking if our Login button changes state or not
it("Login Button Work's",()=>{
    const {getByTestId} = render(<Store><LoginForm></LoginForm></Store>)
    fireEvent.change(getByTestId("username"),{target:{value:'vatsal'}})
    fireEvent.change(getByTestId("password"),{target:{value:'vatsal345'}})
    fireEvent.click(getByTestId("btnLogin"))
})