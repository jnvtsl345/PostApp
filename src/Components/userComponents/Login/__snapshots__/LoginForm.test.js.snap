// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<form
  className="container"
  data-testid="loginForm"
  onSubmit={[Function]}
>
  <label
    htmlFor="username"
  >
    Username:
  </label>
  <input
    data-testid="username"
    id="username"
    name="username"
    onChange={[Function]}
    required={true}
    type="text"
    value=""
  />
  <label
    htmlFor="username"
  >
    Password:
  </label>
  <input
    data-testid="password"
    id="password"
    name="password"
    onChange={[Function]}
    required={true}
    type="password"
    value=""
  />
  <input
    data-testid="btnLogin"
    type="submit"
    value="Login"
  />
</form>
`;
